File: src/types/archive.ts
import { KundaliResponse } from './kundali';
import { WisdomResponse } from './wisdom';
import { RoyalManuscriptData } from './manuscript';
import { MuhuratResponse } from './muhurat';
import { TimelineResponse } from './timeline';

export type ArchiveCategory =
  | 'personal'
  | 'family'
  | 'muhurat'
  | 'manuscript'
  | 'timeline'
  | 'bookmark'
  | 'collection';

export type FamilyGroup =
  | 'self'
  | 'parents'
  | 'children'
  | 'relatives'
  | 'clients'
  | 'custom';

export interface ArchiveMetadata {
  createdAt: string;
  lastViewedAt: string;
  updatedAt: string;
  schemaVersion: string;
  calculationEngineVersion: string;
  appVersion: string;
  ayanamsha: string;
  houseSystem: string;
}

export interface BirthInputData {
  fullName: string;
  dateOfBirth: string;
  timeOfBirth: string;
  city: string;
  latitude?: number;
  longitude?: number;
  timezone?: string;
  gender?: string;
  notes?: string;
}

export interface ArchiveItem {
  id: string;
  title: string;
  category: ArchiveCategory;
  familyGroup: FamilyGroup;
  tags: string[];
  isFavorite: boolean;
  notes?: string;
  birthData: BirthInputData;
  kundaliData?: KundaliResponse;
  wisdomData?: WisdomResponse;
  manuscriptData?: RoyalManuscriptData;
  muhuratData?: MuhuratResponse;
  timelineData?: TimelineResponse;
  metadata: ArchiveMetadata;
}

export interface ArchiveExportData {
  version: string;
  exportedAt: string;
  app: string;
  itemsCount: number;
  items: ArchiveItem[];
  settings?: Record<string, any>;
  checksum?: string;
}

export interface FilterOptions {
  searchQuery: string;
  category: ArchiveCategory | 'all';
  familyGroup: FamilyGroup | 'all';
  rashiFilter: string | 'all';
  nakshatraFilter: string | 'all';
  favoritesOnly: boolean;
  sortBy: 'recent' | 'name' | 'dateOfBirth' | 'lastViewed';
  sortOrder: 'asc' | 'desc';
}

export interface ComparisonMatrixItem {
  id: string;
  title: string;
  familyGroup: FamilyGroup;
  birthData: BirthInputData;
  kundaliData?: KundaliResponse;
}

export interface ChartComparisonResult {
  items: ComparisonMatrixItem[];
  planetMatrix: {
    planet: string;
    placements: Record<string, {
      rashi?: string;
      degree?: string;
      nakshatra?: string;
      house?: number;
      dignity?: string;
    }>;
  }[];
  lagnas: {
    itemId: string;
    title: string;
    rashi: string;
    lord: string;
    nakshatra: string;
    degree: string;
  }[];
  yogas: {
    itemId: string;
    title: string;
    yogaNames: string[];
  }[];
  panchang: {
    itemId: string;
    title: string;
    date?: string;
    time?: string;
    city?: string;
  }[];
}

File: backend/models/archive.py
"""
Pydantic schemas for Phase 14 - Eternal Archive (सनातन अभिलेखागार)
"""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime


class ArchiveMetadataSchema(BaseModel):
    createdAt: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    lastViewedAt: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    updatedAt: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    schemaVersion: str = "1.0.0"
    calculationEngineVersion: str = "1.0.0"
    appVersion: str = "1.0.0"
    ayanamsha: str = "Lahiri"
    houseSystem: str = "Whole Sign"


class ArchiveItemSchema(BaseModel):
    id: str
    title: str
    category: str = Field(..., description="personal | family | muhurat | manuscript | timeline | bookmark | collection")
    familyGroup: str = Field("self", description="self | parents | children | relatives | clients | custom")
    tags: List[str] = Field(default_factory=list)
    isFavorite: bool = False
    notes: Optional[str] = None
    birthData: Dict[str, Any]
    kundaliData: Optional[Dict[str, Any]] = None
    wisdomData: Optional[Dict[str, Any]] = None
    manuscriptData: Optional[Dict[str, Any]] = None
    muhuratData: Optional[Dict[str, Any]] = None
    timelineData: Optional[Dict[str, Any]] = None
    metadata: ArchiveMetadataSchema = Field(default_factory=ArchiveMetadataSchema)


class ArchiveBackupPayload(BaseModel):
    version: str = "1.0.0"
    exportedAt: str = Field(default_factory=lambda: datetime.utcnow().isoformat())
    app: str = "Kalachakra Eternal Archive"
    itemsCount: int
    items: List[ArchiveItemSchema]
    settings: Optional[Dict[str, Any]] = None
    checksum: Optional[str] = None


class ValidationResponse(BaseModel):
    valid: bool
    version: str
    itemCount: int
    errors: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)


class ComparisonRequestSchema(BaseModel):
    itemIds: List[str]
    items: List[ArchiveItemSchema]


class MigrationResultSchema(BaseModel):
    originalVersion: str
    targetVersion: str
    migratedCount: int
    success: bool
    details: List[str] = Field(default_factory=list)
